src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -763,6 +763,78 @@ private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<Attr
                     preserveSig);
             }
         }
+
+        private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is object);
+
+            if (MethodKind != MethodKind.Ordinary)
+            {
+                // Ignore cases where there will already be a warning due to the AttributeTargets.Method usage
+                // restriction on the attribute definition in the framework.
+                if (MethodKind != MethodKind.Constructor && MethodKind != MethodKind.StaticConstructor)
+                {
+                    arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeOrdinary, arguments.AttributeSyntaxOpt.Location);
+                }
+                return;
+            }
+
+            Debug.Assert(ContainingType is object);
+
+            if (isInaccessible(this) || containingTypes(this).Any(isInaccessible))
+            {
+                var topLevelType = containingTypes(this).Last();
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeAccessibleOutsideTopLevelType, arguments.AttributeSyntaxOpt.Location, Name, topLevelType.Name);
+            }
+
+            if (!IsStatic)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeStatic, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (ParameterCount > 0)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotHaveParameters, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (!ReturnsVoid)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustReturnVoid, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (IsGenericMethod)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotBeGeneric, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (ContainingType.IsGenericType)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotBeContainedInGenericType, arguments.AttributeSyntaxOpt.Location, Name);
+            }
```

Comment:
Should I extract the commonalities to a local function or a loop?

Reply:
@RikkiGibson and @cston asked to consider collapsing these, so now there isn't much repetition.